### ISSUE TITLE
Design: what should `domain.shape` do?

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#17909]
Related: `.indices` #15103 #17883; `dims()` method on rectangular domains

Background: `domain.shape` returns a tuple of integers indicating sizes in each dimension. For non-rectangular domains, it is a 1-tuple containing `domain.size`. At the same time we have considered using the term "shape" to describe the underlying index set of an iterator or an iterable expression (when such index set naturally exists).

Consideration: this follows numpy, which is a good precedent. However, if numpy arrays are 0-based, this can be considered an encoding of ranges in each dimension.

This issue proposes:

* `.shape` stays as-is for dense rectangular
  - `.indices` or `.dims()` is/are the things that talk about ranges
  - this supports the rule "if shapes of two domains are equal, they can be assigned/zippered/etc."
* `shape` is not defined for associative domains, as they are not really "shapeful"; so, use `.size` instead
* `shape` on a sparse domain gives same thing as its parent domain
  - the logical shape for a sparse rectangular
  - not defined for a sparse associative
* use `.domain` to get the shape of an iterator/an iterable, by analogy to arrays
  - for some iterators, "shape" is not defined, ex. an iterator performing a recursive traversal of a tree
  - for some iterators, the "shape" is a range, ex. a for-expression over a range
  - so user code might look like:
```chpl
if d.hasDomain() {
  // use d.domain
} else if d.hasRange() {
  // use d.range
} else {
  // do something without relying on the iterator's shape
}
```

**Comment by mppf** [comment]
Generally this approach sounds good to me. I think we should consider a little bit more what alternatives we have for the iterator case to replace `shape`. Do we need to define `hasDomain` and `hasRange` in the near term? Or can we put off this problem because the iterator shape queries are not currently user-facing?

Relatedly I know that in general we like to avoid creating domains when we might not need to. But for the case of iterator domain/ranges specifically, is it low-impact enough on performance to have the iterator create a domain so that the "shape" query can return a domain? Could we put off converting a tuple of ranges into a domain until this query is run? In either case we can have just `someIterator.domain`.

**Comment by bradcray** [comment]
> Or can we put off this problem because the iterator shape queries are not currently user-facing?

I think you're right that we could though, as a module author, I find myself wanting to sometimes do `hasDomain` + `iterRecord.domain` to detect cases that do have them, which might suggest not putting it off too long.  Of course, in order to make them user-facing, we'd probably need to commit to how a user would write a function that took an `_iteratorRecord`, which can still only officially be done by having a completely generic argument.  Interfaces may be part of the solution here, since we could have `Iterable` and `DomainfulIterable` as interfaces that some iterator expressions would meet and others not.

> But for the case of iterator domain/ranges specifically, is it low-impact enough on performance to have the iterator create a domain so that the "shape" query can return a domain? Could we put off converting a tuple of ranges into a domain until this query is run?

I don't feel like I'm understanding this.

**Comment by mppf** [comment]
> > But for the case of iterator domain/ranges specifically, is it low-impact enough on performance to have the iterator create a domain so that the "shape" query can return a domain? Could we put off converting a tuple of ranges into a domain until this query is run?
> 
> I don't feel like I'm understanding this.

@vasslitvinov was saying that iterator "shape" today can be either a range or a domain. Part of the proposal at the top of the issue is that iterators would get `iterator.hasRange` and `iterator.hasDomain` methods and you would use those to decide whether to call `iterator.range` or `iterator.domain`. I was asking if we could make all iterator shapes return a domain as a way of simplifying this situation (because we could remove `iterator.hasRange` and `iterator.range`). I know that creating a domain has overhead but it might be the case that the additional overhead is acceptable in practice.

**Comment by bradcray** [comment]
I see.  My understanding of this feature's role is that it currently serves the role of saying "Can we predict the overall size of this thing?" so that in a case like `var A = [i in myIter()] f(i);` we can know whether we can predict A's size/domain or not.  For example, if I have a `readAllIntsFromFile()` iterator, there's probably no way to know how many things it will return without invoking it, so it would not have the ability to query its range or domain.  I think this is the main reason that we don't always support these queries (i.e., would need a `hasRange` or `hasDomain` or `hasShape` query) rather than concern about domain overhead (?).

**Comment by mppf** [comment]
Yes, that's right. I was proposing we would have `iterator.domain` and `iterator.hasDomain` but that we might not need `iterator.hasRange` or `iterator.range`.

**Comment by bradcray** [comment]
Ah, OK, I guess I still didn't see then.  Interesting question, and I think I agree.  Do you happen to know what cases produce a range, rather than domain, shape today?  Or if not, does @vasslitvinov?

And then as a follow-up question, if I were to do:

```chapel
var A = [i in myIterThatHasRangeShape()] i;
```

would A be an array whose domain was that range value anyway?

**Comment by vasslitvinov** [comment]
> what cases produce a range, rather than domain, shape today?

@bradcray it is when the iterable comes from a range, ex.
```chpl
(forall i in 1..n do i)    // range shape
(forall i in {1..n} do i)  // domain shape

proc foo(i:int, j: int) return i + j;
foo(i, 1..n)    // range shape
foo(i, {1..n})  // domain shape
```

> var A = [i in myIterThatHasRangeShape()] i;

Indeed, A's domain will have that range as its "dimension".

&nbsp;

Overall, I am OK with Michael's suggestion of dropping `iterator.hasRange` / `iterator.range` for now. If we do that, I suggest we leave the door open for introducing them later. For example, call the query `hasDomain` specifically, rather than using a generic name like `hasShape` or `hasIndices`.

There is an interesting question here. With an implicit range-to-domain conversion, should we ensure uniqueness of the implicitly-created domain, ex.:
```chpl
proc test(someIterable) {
  var A: [someIterable.domain] int;
  var B: [someIterable.domain] real;
  var C = someIterable;
}
```

In the above example, if someIterable's shape is a range, should A.domain and B.domain and C.domain have the same identify? Does it matter at all whether they do?

**Comment by bradcray** [comment]
> With an implicit range-to-domain conversion, should we ensure uniqueness of the implicitly-created domain, ex.:

My intuition is that we should not and that it won't matter that we don't (apart from the potential inefficiency of having multiple domains representing the same value).

**Comment by vasslitvinov** [comment]
We are not planning to resolve this issue for 2.0, so I am removing the 2.0 label.
